NIFI-4149 - [WIP] - Indicate if EL is evaluated against FFs or not

pvillard31 opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

pvillard31 commented:
This PR is work in progress.

For now, the idea is to provide a way to specify how is evaluated the expression language: against attributes of incoming flow files (+ registries) or only the variables in registries. The current PR also deprecates the existing ``.expressionLanguageSupported(boolean)`` method.

This PR updates the documentation generation to display the information in the documentation of each processor. Example:
![screenshot-2017-10-10 nifi](https://user-images.githubusercontent.com/11541012/31379809-ac25752c-adaf-11e7-8127-c4ead12e6523.png)

Remaining work:
- update all the existing processors to switch to the correct method / scope
- add this information in the UI (in the tooltip of the properties)

The objective of this current PR is to get feedbacks from the community regarding the current approach before updating all the existing processors.

Thanks!

pvillard31 commented:
Note to myself: improve the documentation to take care of the case exposed in the screenshot. In case the processor does not accept incoming connection, the documentation should not say "evaluated per flow file" even though the scope is saying otherwise. That's because a property can be shared between two processors with different annotations for incoming connections.

joewitt commented on nifi-api/src/main/java/org/apache/nifi/expression/ExpressionLanguageScope.java:
```diff
@@ -0,0 +1,39 @@
+    /**
+     * Expression language is evaluated against variables in registry
+     */
+    VARIABLE_REGISTRY_ONLY,
```
probably makes sense to just name the scope as 'VARIABLE_REGISTRY'

pvillard31 replied:
Done

joewitt commented on nifi-api/src/main/java/org/apache/nifi/expression/ExpressionLanguageScope.java:
```diff
@@ -0,0 +1,39 @@
+package org.apache.nifi.expression;
+
+/**
+ * Indicates the scope of expression language on a property descriptor
```
would be good to add comments here and/or with each scope but they're effectively hierarchical.

NONE -> VARIABLE_REGISTRY -> FLOWFILE

If something supports flowfile it supports both flowfile attributes and variable registry.  If something supports variable registry then it does not consider flowfile attributes.  Further, when supporting the variable registry there is an implied scope of 'variables defined in the closest process group and then up to the higher group and so on until the root group and finally the nifi.properties variables are evaluated and then environment variables/etc..'

pvillard31 replied:
Right, I added some comments to reflect how scope are hierarchically used.

markap14 commented:
@pvillard31 one note to consider is that a Processor is able to also provide its own set of 'variables' to the EL when evaluated. For example, when RouteText is used with a "Matching Strategy" of "Satisfies Expression", the variables "line" and "lineNo" are also available in EL to make routing decisions. I think it would be nice for this to be better reflected here as well, such as :

```
.expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
.expressionLanguageVariable("line", "The line of text to evaluate")
.expressionLanguageVariable("lineNo", "The 1-based number that represents which line in the incoming FlowFile this line is")
```

Additionally, the Mock Framework currently will throw an Exception if a Property indicates that EL is supported but does not evaluate the Expression Language (or vice versa). It probably makes sense to update that to also ensure that the proper scope was used (if VARIABLE_REGISTRY is specified but FlowFile attributes are evaluated, that should fail). This could potentially be done as a separate JIRA, though.

pvillard31 commented:
Thanks for the feedback @markap14! I'll have a look for variables at processor level (never noticed it to be honest). Regarding the mock framework, I added some changes [here](https://github.com/apache/nifi/pull/2205/files#diff-870890bcfd73493b10d56ae9f7e1d3b1R192). Are you expecting something else?

markap14 commented:
@pvillard31 regarding the mock framework: nope, that looks good! I must have just missed it the first time around. Thanks!

pvillard31 commented:
@joewitt @markap14 
I finally found time to get back to this one. I updated every component to use the new method with the EL scope, I updated the mock framwork to fail in case scope is not correctly set, I updated the documentation, changed a bit the UI to add the scope in both the processor documentation and the properties tooltip.

A bit like the recent change with the definition of dependencies in poms, this PR is quite impacting and possibly/probably conflicting with opened PRs. Let me know if you have feedbacks/comments.

Note - @markap14, I didn't address your comment to document the variables made available in some processors like in RouteText. I can do it in this PR or in a follow-up JIRA. Just wanted to get the main work done asap, took me quite some time :)

markap14 commented:
Hey @pvillard31 I definitely where this is going. I agree that another PR/JIRA can be issued for the more elaborate indication of processor-specific variables in the EL. The only issue that I have with this as-is, is that if we have processor that still indicates `.expressionLanguageSupported(true)`, in the UI all that it says is "Expression Language Scope: NONE" and that believes me to believe that Expression Language is not supported. I think if the processor still indicates `true` instead of the Scope, that we still show in the UI: "Expression Language Supported: true" or whatever it is that we show currently.

The only other note, which is very minor, is that in the UI I would avoid showing the Scope as NONE, VARIABLE_REGISTRY, FLOWFILE_ATTRIBUTES and instead use human-friendly syntax: "Not Supported", "Variable Registry Only" and "Variable Registry and FlowFile Attributes" - perhaps just update ExpressionLanguageScope enum to contain a "description" or something and then populate the DTO with that? Thoughts on that?

pvillard31 commented:
@markap14 - thanks for the feedback, completely agree. I pushed a commit that should address your comments and I rebased against master.

markap14 commented:
@pvillard31 thanks for the update! I made a couple of very minor tweaks to verbiage (expanded the term "registry" to "variable registry" in the docs for clarification, and expanded the method name on the dto from "scopeEl" to "ExpressionLanguageScope" for getter/setter). Otherwise, all looks great. Many thanks for tackling this task! It's far from trivial and everything works great! Code looks great, definitely a very welcome improvement to the usability! +1 merged to master.
